File: ja2/src/VideoAddressing.java


public class VideoAddressing
{
	private static final int NTSC_LINES_PER_FRAME = 3*5*5*7;
	private static final int NTSC_LINES_PER_FIELD = NTSC_LINES_PER_FRAME/2;
	private static final int NTSC_FIELDS_PER_SECOND = 60;
	private static final int NTSC_COLOR_FIELD_EVERY = 1000;

	private static final int CRYSTAL_HZ = divideRoundUp(315000000,22);
	private static final int CPU_CYCLES_PER_CRYSTAL_CYCLES = 14;
	private static final int CPU_HZ = divideRoundUp(CRYSTAL_HZ,CPU_CYCLES_PER_CRYSTAL_CYCLES);
	private static final int APPLE_BYTES = (NTSC_COLOR_FIELD_EVERY+1)*CPU_HZ;
	private static final int LINES = NTSC_FIELDS_PER_SECOND*NTSC_COLOR_FIELD_EVERY*NTSC_LINES_PER_FIELD;
	private static final int BYTES_PER_ROW = APPLE_BYTES/LINES;

	public static final int BYTES_PER_FIELD = BYTES_PER_ROW*NTSC_LINES_PER_FIELD;

	private static final int VISIBLE_BYTES_PER_ROW = 40;
	private static final int BLANKED_BYTES_PER_ROW = BYTES_PER_ROW-VISIBLE_BYTES_PER_ROW;
	private static final int VISIBLE_ROWS_PER_FIELD = 192;
	private static final int VISIBLE_BYTES_PER_FIELD = BYTES_PER_ROW*VISIBLE_ROWS_PER_FIELD;
	private static final int SCANNABLE_ROWS = 0x100;
	private static final int SCANNABLE_BYTES = SCANNABLE_ROWS*BYTES_PER_ROW;
	private static final int RESET_ROWS = NTSC_LINES_PER_FIELD-SCANNABLE_ROWS;
	private static final int RESET_BYTES = RESET_ROWS*BYTES_PER_ROW;

	public static int[] buildLUT(final int base, final int len)
	{
		final int[] lut = new int[BYTES_PER_FIELD];
		for (int t = 0; t < lut.length; ++t)
		{
			lut[t] = base + (calc(t) % len);
		}
		return lut;
	}

	private static int divideRoundUp(final int dividend, final int divisor)
	{
		return (dividend+divisor-1)/divisor;
	}

	private static int calc(final int t)
	{
		int c = t % VISIBLE_BYTES_PER_FIELD;
		if (t >= SCANNABLE_BYTES)
		{
			c -= RESET_BYTES;
		}

		int n = c / BYTES_PER_ROW;
		final int s = (n/0x40);
		n -= s*0x40;
		final int q = (n/8);
		n -= q*8;
		final int base = 0x400*n+0x80*q+VISIBLE_BYTES_PER_ROW*s;

		final int half_page = (base >> 7) << 7;

		int a = base+(c%BYTES_PER_ROW)-BLANKED_BYTES_PER_ROW;
		if (a < half_page)
		{
			a += 0x80;
		}
		return a;
	}
}

File: epple2sdl/src/videoaddressing.cpp
/*
    epple2
    Copyright (C) 2008 by Chris Mosher <chris@mosher.mine.nu>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY, without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
#include "timinggenerator.h"
#include "videoaddressing.h"

VideoAddressing::VideoAddressing()
{
}
static int calc(const unsigned int t)
{
	int c = t % VideoAddressing::VISIBLE_BYTES_PER_FIELD;
	if (t >= VideoAddressing::SCANNABLE_BYTES)
	{
		c -= VideoAddressing::RESET_BYTES;
	}

	int n = c / VideoAddressing::BYTES_PER_ROW;
	const int s = (n >> 6);
	n -= s << 6;
	const int q = (n >> 3);
	n -= q << 3;
	const int base = (n<<10) + (q<<7) + VideoAddressing::VISIBLE_BYTES_PER_ROW*s;

	const int half_page = base & 0xFF80;

	int a = base+(c%VideoAddressing::BYTES_PER_ROW)-VideoAddressing::BLANKED_BYTES_PER_ROW;
	if (a < half_page)
	{
		a += 0x80;
	}
	return a;
}

void VideoAddressing::buildLUT(const unsigned short base, const unsigned short len, std::vector<unsigned short>& lut)
{
	lut.resize(BYTES_PER_FIELD);
	for (unsigned int t = 0; t < BYTES_PER_FIELD; ++t)
	{
		unsigned int off = (calc(t) % len);

		const unsigned int col = t % BYTES_PER_ROW;
		const unsigned int row = t / BYTES_PER_ROW;

		if (col < BLANKED_BYTES_PER_ROW)
		{
			// HBL
			if (base < 0x1000)
			{
				off += 0x1000;
			}
			if (col == 0)
			{
				++off;
			}
		}

		if (row >= VISIBLE_ROWS_PER_FIELD)
		{
			// VBL
			const int base2 = off & 0xFF80;
			off &= 0x7F;
			off -= 8;
			off &= 0x7F;
			off += base2;
		}

		lut[t] = base + off;
	}
}
